// import-sort-config/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assign = require("core-js/library/fn/object/assign");
const minimatch = require("minimatch");
const fs_1 = require("fs");
const findRoot = require("find-root");
const resolve = require("resolve-from").silent;
;
exports.DEFAULT_CONFIGS = {
    ".js, .jsx, .es6, .es": {
        parser: "babylon",
        style: "eslint",
    },
    ".ts, .tsx": {
        parser: "typescript",
        style: "eslint",
    },
};
function getConfig(extension, directory) {
    const defaultConfig = getConfigForExtension(exports.DEFAULT_CONFIGS, extension);
    let packageConfig;
    if (directory) {
        packageConfig = getConfigFromPackage(directory, extension);
    }
    const actualConfig = mergeConfigs([defaultConfig, packageConfig]);
    if (!actualConfig) {
        return;
    }
    return resolveConfig(actualConfig, directory);
}
exports.getConfig = getConfig;
function getConfigFromPackage(directory, extension) {
    const packageConfigs = getConfigsFromPackage(directory);
    if (!packageConfigs) {
        return;
    }
    return getConfigForExtension(packageConfigs, extension);
}
function getConfigForExtension(configs, extension) {
    const foundConfigs = Object.keys(configs).map(joinedGlobs => {
        const globs = joinedGlobs.split(",").map(rawGlob => rawGlob.trim());
        const config = configs[joinedGlobs];
        if (globs.some(glob => minimatch(extension, glob))) {
            return config;
        }
    });
    return mergeConfigs(foundConfigs);
}
function getConfigsFromPackage(directory) {
    try {
        return JSON.parse(fs_1.readFileSync(`${findRoot(directory)}/package.json`, "utf-8")).importSort;
    }
    catch (e) {
        return;
    }
}
function mergeConfigs(rawConfigs) {
    const configs = rawConfigs.filter(rawConfig => !!rawConfig);
    if (configs.length === 0) {
        return;
    }
    return configs.reduce((previousConfig, currentConfig) => {
        if (!currentConfig) {
            return previousConfig;
        }
        const config = assign({}, previousConfig);
        if (currentConfig.parser) {
            config.parser = currentConfig.parser;
        }
        if (currentConfig.style) {
            config.style = currentConfig.style;
        }
        return config;
    });
}
function resolveConfig(config, directory) {
    const resolvedConfig = {
        config,
    };
    if (config.parser) {
        resolvedConfig.parser = resolveParser(config.parser, directory);
    }
    if (config.style) {
        resolvedConfig.style = resolveStyle(config.style, directory);
    }
    return resolvedConfig;
}
function resolveParser(module, directory) {
    if (module.indexOf(".") === 0) {
        return resolveModule(module, directory);
    }
    if (module.indexOf("import-sort-parser") === 0) {
        return resolveModule(module, directory);
    }
    return resolveModule(`import-sort-parser-${module}`, directory);
}
function resolveStyle(module, directory) {
    if (module.indexOf(".") === 0) {
        return resolveModule(module, directory);
    }
    if (module.indexOf("import-sort-style-") === 0) {
        return resolveModule(module, directory);
    }
    return resolveModule(`import-sort-style-${module}`, directory);
}
function resolveModule(module, directory) {
    if (directory) {
        const path = resolve(directory, module);
        if (path) {
            return path;
        }
    }
    const path = resolve(__dirname, module);
    if (path) {
        return path;
    }
}
