// stylelint/dist/rules/font-weight-notation/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.messages = exports.ruleName = undefined;

exports.default = function (expectation, options) {
  return function (root, result) {
    var validOptions = (0, _utils.validateOptions)(result, ruleName, {
      actual: expectation,
      possible: ["numeric", "named"]
    }, {
      actual: options,
      possible: {
        ignore: ["relative"]
      },
      optional: true
    });
    if (!validOptions) {
      return;
    }

    root.walkDecls(function (decl) {
      if (decl.prop === "font-weight") {
        checkWeight(decl.value, decl);
      }

      if (decl.prop === "font") {
        checkFont(decl);
      }
    });

    function checkFont(decl) {
      var _iteratorNormalCompletion = true;
      var _didIteratorError = false;
      var _iteratorError = undefined;

      try {
        for (var _iterator = _postcss2.default.list.space(decl.value)[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
          var value = _step.value;

          // We do not need to more carefully distinguish font-weight
          // numbers from unitless line-heights because line-heights in
          // `font` values need to be part of a font-size/line-height pair
          if (isNumbery(value) || NAMED_WEIGHTS.indexOf(value) !== -1) {
            checkWeight(value, decl);
            return;
          }
        }
      } catch (err) {
        _didIteratorError = true;
        _iteratorError = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion && _iterator.return) {
            _iterator.return();
          }
        } finally {
          if (_didIteratorError) {
            throw _iteratorError;
          }
        }
      }
    }

    function checkWeight(weightValue, decl) {
      if ((0, _utils.cssWordIsVariable)(weightValue)) {
        return;
      }
      if ((0, _utils.optionsHaveIgnored)(options, "relative") && RELATIVE_NAMED_WEIGHTS.indexOf(weightValue) !== -1) {
        return;
      }

      var weightValueOffset = decl.value.indexOf(weightValue);

      if (expectation === "numeric") {
        if (!isNumbery(weightValue)) {
          return complain(messages.expected("numeric"));
        }
      }

      if (expectation === "named") {
        if (isNumbery(weightValue)) {
          return complain(messages.expected("named"));
        }
        if (NAMED_WEIGHTS.indexOf(weightValue) === -1) {
          return complain(messages.invalidNamed(weightValue));
        }
        return;
      }

      function complain(message) {
        (0, _utils.report)({
          ruleName: ruleName,
          result: result,
          message: message,
          node: decl,
          index: (0, _utils.declarationValueIndexOffset)(decl) + weightValueOffset
        });
      }
    }
  };
};

var _postcss = require("postcss");

var _postcss2 = _interopRequireDefault(_postcss);

var _utils = require("../../utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ruleName = exports.ruleName = "font-weight-notation";

var messages = exports.messages = (0, _utils.ruleMessages)(ruleName, {
  expected: function expected(type) {
    return "Expected " + type + " font-weight notation";
  },
  invalidNamed: function invalidNamed(name) {
    return "Unexpected invalid font-weight name \"" + name + "\"";
  }
});

var NAMED_WEIGHTS = ["normal", "bold", "bolder", "lighter"];
var RELATIVE_NAMED_WEIGHTS = ["bolder", "lighter"];
function isNumbery(x) {
  return Number(x) == x;
}
